File: 2375_construct_smallest_number_from_di_string.cpp
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    string smallestNumber(string f) {
        string g;
        int h = f.length();
        vector<int> i(h, 0); 
        i[h - 1] = (f[h - 1] == 'D') ? 1 : 0;
        for (int j = h - 2; j >= 0; --j) 
            if (f[j] == 'D') i[j] = i[j + 1] + 1;
        int k = 1;
        for (int j = 0; j < h; ++j) {
            int l = i[j] + k;
            if (l == k) k = j + 2;
            g += to_string(l);
        }
        return g + to_string(k);
    }
};

File: 2375_construct_smallest_number_from_di_string.java
class Solution {
    public String smallestNumber(String m) {
        StringBuilder z = new StringBuilder();
        int x = m.length(), q[] = new int[x];
        q[x - 1] = (m.charAt(x - 1) == 'D') ? 1 : 0;
        for (int i = x - 2; i >= 0; --i) if (m.charAt(i) == 'D') q[i] = q[i + 1] + 1;
        int y = 1;
        for (int i = 0; i < x; ++i) {
            int w = q[i] + y;
            if (w == y) y = i + 2;
            z.append(w);
        }
        return z.append(y).toString();
    }
}
